### ISSUE TITLE
Unify how lamp.steps stores data across device types.

### ISSUE THREAD

**Issue by lukeoftheshire** [BIDMCDigitalPsychiatry/LAMP-platform#503]
The `lamp.steps` sensor implementation is not consistent across device types.

Example Android data:
```
U6840262928 {'data': [{'data': {'type': 'speed', 'unit': 'meters per second', 'value': 0.99644613}, 'sensor': 'lamp.steps', 'timestamp': 1644331154982},
                      {'data': {'source': 'com.google.android.gms', 'type': 'active_minutes', 'value': 1}, 'sensor': 'lamp.steps', 'timestamp': 1644331154982},
                      {'data': {'type': 'distance_delta', 'unit': 'meter', 'value': 25.457724}, 'sensor': 'lamp.steps', 'timestamp': 1644331154981},
                      {'data': {'source': 'com.google.android.gms', 'steps': 38, 'type': 'step_count_delta'}, 'sensor': 'lamp.steps', 'timestamp': 1644331154955}, 
                      {'data': {'source': 'com.google.android.gms', 'steps': 29, 'type': 'step_count_delta'}, 'sensor': 'lamp.steps', 'timestamp': 1644331154953}, 
                      {'data': {'source': 'com.google.android.gms', 'steps': 13, 'type': 'step_count_delta'}, 'sensor': 'lamp.steps', 'timestamp': 1644331154951}, 
                      {'data': {'source': 'com.google.android.gms', 'steps': 10, 'type': 'step_count_delta'}, 'sensor': 'lamp.steps', 'timestamp': 1644331154948}, 
                      {'data': {'source': 'com.google.android.gms', 'steps': 15, 'type': 'step_count_delta'}, 'sensor': 'lamp.steps', 'timestamp': 1644331154945}, 
                      {'data': {'source': 'com.google.android.gms', 'steps': 13, 'type': 'step_count_delta'}, 'sensor': 'lamp.steps', 'timestamp': 1644331154942},
                      {'data': {'source': 'com.google.android.gms', 'steps': 9, 'type': 'step_count_delta'}, 'sensor': 'lamp.steps', 'timestamp': 1644329960659}]} 

```

Example iOS data:
```
U4842964680 {'data': [{'sensor': 'lamp.steps', 'data': {'value': 18, 'unit': 'count', 'source': 'com.apple.health.CDD08AB1-5DEE-449C-B664-9EBCBA6D0BF7'}, 'timestamp': 1644303550327},
                      {'sensor': 'lamp.steps', 'data': {'value': 445, 'unit': 'count', 'source': 'com.apple.health.CDD08AB1-5DEE-449C-B664-9EBCBA6D0BF7'}, 'timestamp': 1644303524689},
                      {'sensor': 'lamp.steps', 'data': {'value': 325, 'unit': 'count', 'source': 'com.apple.health.DF1B23E1-916F-4C4B-8A76-8AFC03DA3E22'}, 'timestamp': 1644303518522},
                      {'sensor': 'lamp.steps', 'data': {'distance': 436.66149301594123, 'floors_descended': 0, 'event': 'pause', 'active_pace': 0.8171822726694453, 'floors_ascended': 4, 'value': 550, 'pace': 0, 'cadence': 0}, 'timestamp': 1644303517293},
                      {'sensor': 'lamp.steps', 'data': {'distance': 422.3853307836689, 'floors_descended': 0, 'event': 'resume', 'active_pace': 0.8202947486532274, 'floors_ascended': 4, 'value': 530, 'pace': 0, 'cadence': 0}, 'timestamp': 1644303503122}, 
                      {'sensor': 'lamp.steps', 'data': {'distance': 418.67355938861147, 'floors_descended': 0, 'event': 'pause', 'active_pace': 0.8199164624852057, 'floors_ascended': 4, 'value': 523, 'pace': 0, 'cadence': 0}, 'timestamp': 1644303478620},
                      {'sensor': 'lamp.steps', 'data': {'distance': 414.55335054174066, 'floors_descended': 0, 'event': 'resume', 'active_pace': 0.8186427472037264, 'floors_ascended': 4, 'value': 517, 'pace': 0, 'cadence': 0}, 'timestamp': 1644303476041},
                      {'sensor': 'lamp.steps', 'data': {'distance': 414.55335054174066, 'floors_descended': 0, 'event': 'pause', 'active_pace': 0.815099777407244, 'floors_ascended': 4, 'value': 517, 'pace': 0, 'cadence': 0}, 'timestamp': 1644303465728}, 
                      {'sensor': 'lamp.steps', 'data': {'distance': 408.9478182308376, 'floors_descended': 0, 'event': 'resume', 'active_pace': 0.8141033315455283, 'floors_ascended': 4, 'value': 509, 'pace': 0, 'cadence': 0}, 'timestamp': 1644303460571}, 
                      {'sensor': 'lamp.steps', 'data': {'distance': 405.14135878998786, 'floors_descended': 0, 'event': 'pause', 'active_pace': 0.8148390502168555, 'floors_ascended': 4, 'value': 504, 'pace': 0, 'cadence': 0}, 'timestamp': 1644303455416}]} 

```

In my opinion there are (at least) two problems demonstrated here:

1. What is stored as sensor data should not differ across devices as this will complicate pipelines that analyze step data, like LAMP-cortex.
2. One sensor should have a single format of data - more specifically, keys should not appear in one SensorEvent and not another (e.g. how 'distance' and 'unit' behave in iOS data)

W.r.t issue number `1` I think we should come up with a standard format that LAMP tranforms data into before it is sent to the server. For number `2` I think we should perhaps consider breaking out additional information that we don't currently use (but could want later), such as floors ascended/descended or distance traveled into a separate lamp sensor, and not storing it in `lamp.steps`.

**Comment by avaidyam** [comment]
[Another very relevant concern is that the Android telephony sensor doesn't seem to be working at all.](https://mindlamp.discourse.group/t/no-text-call-sensor-data-for-androids/532) This issue also isn't a `feature`, but rather a `bug`, as all sensor data formats between iOS and Android *must* be unified.

**Comment by avaidyam** [comment]
@jijopulikkottil @saranya-sajeev Can you share more details on why the data format is different between Android and iOS at the moment? //cc @ZCOEngineer

**Comment by dcurrey88** [comment]
It also seems that some Androids are not collecting step data. Of the 23 participants with Androids in one of our studies run this past December (researcher id: 4aq1kry81ktrb5v1smvs) only 2 of them have collected any lamp.steps data. All of the iOS devices in the study collected step data.

**Comment by lukeoftheshire** [comment]
Also, not every event has a 'source' key/value pair, so it's unclear what its presence signifies.

**Comment by avaidyam** [comment]
@dcurrey88 Can you check their latest `lamp.analytics` events to see what the `user_agent` strings are? It should also have the `NativeCore` version info in that - which we can use to check which app version the 2 participants with working step count data are using. 

@saranya-sajeev @ZCOEngineer @michaelmenon This issue is now being marked urgent.

**Comment by dcurrey88** [comment]
> @dcurrey88 Can you check their latest `lamp.analytics` events to see what the `user_agent` strings are? It should also have the `NativeCore` version info in that - which we can use to check which app version the 2 participants with working step count data are using.
> 
> @saranya-sajeev @ZCOEngineer @michaelmenon This issue is now being marked urgent.

Here are the `user_agent` strings for the two users with some step data:
- 7888514,Google,Pixel 4a
- G960USQS9FUG2,samsung,SM-G960U

Here are the other `user_agent` strings:
- 729b7f,motorola,moto g(7) power
- d00bb5,motorola,moto g(7)
- 7807550,Google,Pixel 4a (5G)
- G988U1UES2DUJ7,samsung,SM-G988U1
- G970U1UEU6GUJ3,samsung,SM-G970U1
- G991USQU4BUKF,samsung,SM-G991U
- 5ddbb4,motorola,moto g stylus
- N976USQS5EUJ7,samsung,SM-N976U
- 7888514,Google,Pixel 3a XL
- G973USQU6GUJ3,samsung,SM-G973U
- G960USQS9FUG2,samsung,SM-G960U
- G996USQU4BUK9,samsung,SM-G996U
- G998USQU4BUKF,samsung,SM-G998U
- F926U1UES1AUJ7,samsung,SM-F926U1
- 2108161852,OnePlus,GM1917
- 2109301204,OnePlus,KB2007
- A326USQU6AUJ4,samsung,SM-A326U
- 2109301920,OnePlus,IN2017
- 7641976,Google,Pixel 3a
- 2010292059,OnePlus,ONEPLUS A5010
- G973U1UES6GUK1,samsung,SM-G973U1

**Comment by avaidyam** [comment]
@dcurrey88 Do the analytics events have any other strings in them? There should be one with app versions (dashboard, backend, native core) too.

**Comment by avaidyam** [comment]
> `NativeCore 2021.10.20; iOS 15.1; iPhone iPhone10,4` 

This looks like the version reporting is only implemented on iOS and not Android. @dcurrey88 Can you park a new GitHub issue for this?

**Comment by ZCOEngineer** [comment]
> @jijopulikkottil @saranya-sajeev Can you share more details on why the data format is different between Android and iOS at the moment? //cc @ZCOEngineer

@avaidyam we will be checking this today

**Comment by jijopulikkottil** [comment]
The main difference with Android is, in android it is  "steps" instead of "value". We can keep "value" in all devices.
So data will look like:
In Android -> `{ 'sensor': 'lamp.steps', 'data': {'value': 9, 'source': 'com.google.android.gms', 'type': 'step_count_delta'}, 'timestamp': 1644329960659}]} `
In iOS.      ->    `{'sensor': 'lamp.steps', 'data': {'value': 445, 'source': 'com.apple.health.CDD08AB1-5DEE-449C-B664-9EBCBA6D0BF7' 'unit': 'count'}, 'timestamp': 1644303524689}`

The difference in above data is, Android have a 'type' field and iOS have 'unit'. In iOS, we are keeping 'unit' in all sensors from Health Kit.          



In Android, we are also reporting `speed`,` active-minutes`, `distance` etc under` lamp.steps`. Let us know if we need to remove all these from `lamp.steps`.

In iOS, when collect steps from **pedometer**, we are also getting `floorsAscended`, `distance` etc. and reporting the same with step count.

Please suggest if need any changes.

**Comment by saranya-sajeev** [comment]
> [Another very relevant concern is that the Android telephony sensor doesn't seem to be working at all.](https://mindlamp.discourse.group/t/no-text-call-sensor-data-for-androids/532) This issue also isn't a `feature`, but rather a `bug`, as all sensor data formats between iOS and Android _must_ be unified.

Telephony sensor is not implemented in Android. We will integrate it asap.

**Comment by saranya-sajeev** [comment]
> > `NativeCore 2021.10.20; iOS 15.1; iPhone iPhone10,4`
> 
> This looks like the version reporting is only implemented on iOS and not Android. @dcurrey88 Can you park a new GitHub issue for this?
> 
> > `LAMP-dashboard/2021.11.11`
> 
> This is concerning as we've had a number of releases since Nov. 11th, 2021, the latest of which was Jan 11th, 2022. I'm now concerned that the on-device dashboard is not being updated correctly, which could lead to **many** of the issues we've been seeing in the last few months.
> 
> Can you park a new issue for this as well, @dcurrey88? Thanks!

@avaidyam For Login and Log out, there is a mismatch in sending user_agent value. We will make it uniform.

**Comment by avaidyam** [comment]
@saranya-sajeev Can you elaborate **specifically** on how you are going to make it uniform? (i.e. which events will change and what user agent format is going to be used?)

**Comment by saranya-sajeev** [comment]
@avaidyam For low power mode and diagnostic we are sending user_agent  as" Nativecore app version name; device manufacturer; device model" . We will implement that for login and log out too.

**Comment by jijopulikkottil** [comment]
For `lamp.steps`, **iOS** and **Android** can keep the same format below:
`{ 'sensor': 'lamp.steps', 'data': {'value': 9, 'source': 'com.google.android.gms', 'unit': 'count'}, 'timestamp': 1644329960659}]}`
But from iOS, when post lamp.steps from pedometer, the `source` field will be null. (but from HealthKit, there will be `source`)  hope this is okay, or should we add any static text?

If we need to report other additional pedometer properties as different event, two options:
1. we can keep same sensor identifier "lamp.steps" for all and add 'type' parameter. So above data will be like:
`{ 'sensor': 'lamp.steps', 'data': {'type': 'step_count_delta', 'value': 9, 'source': 'com.google.android.gms', 'unit': 'count'}, 'timestamp': 1644329960659}]}`

2. We can use 'lamp.steps' only for step count. And for all others we can use new identifier `lamp.steps_activity` or `lamp.activity` (FYI sensor identifier `lamp.activity_recognition` is already exist)
So lamp.steps will be like:
`{ 'sensor': 'lamp.steps', 'data': {'value': 9, 'source': 'com.google.android.gms', 'unit': 'count'}, 'timestamp': 1644329960659}]}`
And for 'lamp.steps_activity, we can add `type`
`{ 'sensor': 'lamp.steps_activity', 'data': {'type': 'distance', 'value': 9}, 'timestamp': 1644329960659}]}`
for 'lamp.steps_activity, iOS doesn't have `source` or `unit` fields. But Android have. So do we need to remove those two keys? or Should we add static texts when post from iOS?

------
**The possible 'type' from Android are:**
speed, 
active-minutes, 
distance, 
cycling_wheel_cumulative, 
cycling_pedal_cadence, 
heart_minutes, 
power, 
step_count_cadence, 
cycling_wheel_rpm

**from iOS are:**
frequencySpeed
distance
currentPace
currentCadence
floorsAscended
floorsDescended
averageActivePace
event -> (resume or pause)

We assume, there are 3 common types: 'speed', distance', and 'cadence'. Also need to double check these three are return same expected value in both platforms.
 
So should we use all available types ? or use only common 3 ?

**Comment by ZCOEngineer** [comment]
> For `lamp.steps`, **iOS** and **Android** can keep the same format below: `{ 'sensor': 'lamp.steps', 'data': {'value': 9, 'source': 'com.google.android.gms', 'unit': 'count'}, 'timestamp': 1644329960659}]}` But from iOS, when post lamp.steps from pedometer, the `source` field will be null. (but from HealthKit, there will be `source`) hope this is okay, or should we add any static text?
> 
> If we need to report other additional pedometer properties as different event, two options:
> 
> 1. we can keep same sensor identifier "lamp.steps" for all and add 'type' parameter. So above data will be like:
>    `{ 'sensor': 'lamp.steps', 'data': {'type': 'step_count_delta', 'value': 9, 'source': 'com.google.android.gms', 'unit': 'count'}, 'timestamp': 1644329960659}]}`
> 2. We can use 'lamp.steps' only for step count. And for all others we can use new identifier `lamp.steps_activity` or `lamp.activity` (FYI sensor identifier `lamp.activity_recognition` is already exist)
>    So lamp.steps will be like:
>    `{ 'sensor': 'lamp.steps', 'data': {'value': 9, 'source': 'com.google.android.gms', 'unit': 'count'}, 'timestamp': 1644329960659}]}`
>    And for 'lamp.steps_activity, we can add `type`
>    `{ 'sensor': 'lamp.steps_activity', 'data': {'type': 'distance', 'value': 9}, 'timestamp': 1644329960659}]}`
>    for 'lamp.steps_activity, iOS doesn't have `source` or `unit` fields. But Android have. So do we need to remove those two keys? or Should we add static texts when post from iOS?
> 
> **The possible 'type' from Android are:** speed, active-minutes, distance, cycling_wheel_cumulative, cycling_pedal_cadence, heart_minutes, power, step_count_cadence, cycling_wheel_rpm
> 
> **from iOS are:** frequencySpeed distance currentPace currentCadence floorsAscended floorsDescended averageActivePace event -> (resume or pause)
> 
> We assume, there are 3 common types: 'speed', distance', and 'cadence'. Also need to double check these three are return same expected value in both platforms.
> 
> So should we use all available types ? or use only common 3 ?

@avaidyam could you share your thoughts on this here?

**Comment by lukeoftheshire** [comment]
I don't have any issue with having a null value in the `source` field. My preference would be to never have keys present in one situation but entirely absent in another.

> `lamp.activity_recognition`

This is mentioned in the docs but with no actual documentation of what it is/does - if this is an implemented sensor with a viable use case that should change. (https://docs.lamp.digital/data_science/data_types#sensor-types)

I like the idea of breaking out the data into `lamp.steps` and `lamp.steps_activity`. Again, my preference would be to include the three common fields and not the others -- but I can see how having the other data might be useful. As long as the three common fields have the same keys I would be happy.

**Comment by avaidyam** [comment]
> I like the idea of breaking out the data into `lamp.steps` and `lamp.steps_activity`. 

Can you elaborate on what you mean with this? Which of the `type`s would belong to which?

**Comment by lukeoftheshire** [comment]
Yeah, sure! This is in reference to 

>We can use 'lamp.steps' only for step count. And for all others we can use new identifier lamp.steps_activity or lamp.activity

from @jijopulikkottil  a couple comments up.

My understanding is that `lamp.steps` would become what we use for *step count only* where something like `lamp.steps_activity` would contain info like 'cadence' or 'speed' or 'distance' - which are of course useful but might fit better under a non `lamp.steps` sensor. (Please correct me if I am misunderstanding, @jijopulikkottil )

**Comment by avaidyam** [comment]
@lukeoftheshire The problem is, all of those `type`s are [actually part of the pedometer data itself](https://developer.apple.com/documentation/coremotion/cmpedometerdata), on both iOS and Android. I don't think it makes sense to separate the data like that into two related but unlinked sensors...  I like Option 1 that @jijopulikkottil proposed: 

```json
{
    "timestamp": 1644329960659,
    "sensor": "lamp.steps", 
    "data": {
        "value": 9, 
        "unit": "count",
        "type": "step_count_delta", 
        "source": "com.google.android.gms"
    }
}
```

But we do need to unify the `type` parameter such that: 
- We merge iOS and Android terms like so: 
    - `speed == frequencySpeed` → `speed`
    - `distance == distance` → `distance`
    - `step_count_cadence == currentCadence` → `cadence`
- Only iOS offers `currentPace` and `averageActivePace` (→ `pace`?), which are useful for steps, so we should see if Google Fit has equivalent types. 
- Google Fit should also have a `floorsAscended` and `floorsDescended` (→ `floors_ascended` and `floors_descended`?) equivalent, can we check that, @jijopulikkottil?
- Both `heart-minutes` and `active-minutes` are entirely a Google Fit thing that don't matter at all, so we need to remove those. 
- `cycling_wheel_cumulative`, `cycling_pedal_cadence`, and `cycling_wheel_rpm` seem to be unrelated to steps so we can remove those too. 
    - I'm not sure what `power` is but it might be related to cycling. 
    - On iOS, `event.resume/pause` is likely when the individual stops walking and resumes walking, so we don't need that info either.

**Comment by saranya-sajeev** [comment]
@avaidyam Google fit not providing following types
 floorsAscended and  floorsDescended , 
currentPace and averageActivePace,

**Comment by jijopulikkottil** [comment]
@avaidyam floorsAscended, floorsDescended, currentPace and averageActivePace are entirely iOS thing. So do we need those types ?

**Comment by jijopulikkottil** [comment]
By mistake I added 'speed', in iOS, speed is also unavailable from pedometer data.

In iOS,   when an event happens, the 'lamp.steps' sensor record from pedometer is like:

`{'sensor': 'lamp.steps', 'data': {'distance': 418.67355938861147, 'floors_descended': 0, 'event': 'pause', 'active_pace': 0.8199164624852057, 'floors_ascended': 4, 'value': 523, 'pace': 0, 'cadence': 0}, 'timestamp': 1644303478620}`

After our implementation,It will be 2 or 3 records:
```
{
    "timestamp": 1644303478620,
    "sensor": "lamp.steps", 
    "data": {
        "value": 523, 
        "type": "step_count"
    }
}

{
    "timestamp": 1644303478620,
    "sensor": "lamp.steps", 
    "data": {
        "value": 418.67355938861147, 
        "type": "distance"
    }
}

{
    "timestamp": 1644303478620,
    "sensor": "lamp.steps", 
    "data": {
        "value": 0, 
        "type": "cadence"
    }
}

```
@avaidyam  Do we need to create record if 'cadence' value is zero ?

**Comment by avaidyam** [comment]
> floorsAscended, floorsDescended, currentPace and averageActivePace are entirely iOS thing. So do we need those types ?

Let's keep them, but make sure they're converted to `snake_case` (i.e. `average_active_pace` instead of `averageActivePace`). My team will document that these specific sub-types of `lamp.steps` are iOS only for now (though Google Fit does seem to support some of them, just not via the API). //cc @lukeoftheshire @akmeyer3 @dcurrey88 

> Do we need to create record if 'cadence' value is zero ?

Is it `0` because the original data source had `nil` values? If so, ignore it. If the data specifically marked it as `0`, then we should create the record.

**Comment by jijopulikkottil** [comment]
The `type`  values available are:

Below are common for iOS and Android:

```
step_count
distance
cadence
```

Below are only available in iOS:
```
floors_ascended
floors_descended
average_active_pace
pace
```

Below are only available in Android:
`speed`

Below are the iOS sample data:
```
 {
            "sensor": "lamp.steps",
            "data": {
                "value": 3047,
                "type": "step_count"
            },
            "timestamp": 1646581010984
},
{
            "sensor": "lamp.steps",
            "data": {
                "value": 1817.3491907184944,
                "type": "distance"
            },
            "timestamp": 1646581003467
},
{
            "sensor": "lamp.steps",
            "data": {
                "value": 7,
                "type": "floors_descended"
            },
            "timestamp": 1646581003467
},
{
            "sensor": "lamp.steps",
            "data": {
                "value": 127,
                "type": "step_count"
                "unit": "count",
                "source": "com.apple.health.BA2F6275-6CA9-4BC9-8990-D9BC027A3BAF"
            },
            "timestamp": 1646573476438
}
```

Below are the Android sample data:
```
{
      "data": {
        "type": "speed",
        "unit": "meters per second",
        "value": 0.30389535
      },
      "sensor": "lamp.steps",
      "timestamp": 1646643039075
},
{
      "data": {
        "type": "distance",
        "unit": "meter",
        "value": 18.233723
      },
      "sensor": "lamp.steps",
      "timestamp": 1646642993303
}
```


Note - `source` will be available when fetch data from Health Kit and sometimes available from GoogleFit.

**Comment by avaidyam** [comment]
@jijopulikkottil Is there any benefit to the `unit` field here on iOS? What other possible values are there? It needs to be present for **every** event. 

```json
"type": "step_count"
"unit": "count",
```

And for Android, @saranya-sajeev, where are these values coming from?

```json
"unit": "meters per second",
"unit": "meter",
```

Otherwise, looks good.

**Comment by jijopulikkottil** [comment]
@avaidyam 
In iOS, we are getting `unit` field when fetch data from HealthKit. 
Once we had removed `unit` field from all sensor data and reinstated again.

In Android `unit` field value is hardcoded for GoogleFit sensor data.

Let us know if we need to keep `unit` field.

**Comment by avaidyam** [comment]
@jijopulikkottil We need to keep the `unit` field consistent for all `lamp.steps` events, even if its value is `null`.

**Comment by jijopulikkottil** [comment]
@avaidyam ok, we can keep `unit` field for all `lamp.steps` events. 
FYI - The type `cadence` have different unit in iOS and Android. In iOS it is `steps per second`, but for android it is `steps per minute`

**Comment by avaidyam** [comment]
@jijopulikkottil @saranya-sajeev If the `source` field is not available, it **MUST** be set to `null` in the JSON output. If `cadence` has different units, we should convert it, specifically, @jijopulikkottil, please convert it to `steps per minute` on iOS (by multiplying `x 60`).

**Comment by saranya-sajeev** [comment]
@avaidyam Android part is completed, we can provide to QA once iOS is done.

**Comment by jijopulikkottil** [comment]
Test case: in both iOS and Android, for all ` lamp.steps` event, there will be four fields[value, type, uni, source] in the `data` block.

```
{
            "sensor": "lamp.steps",
            "data": {
                "value": 127,
                "type": "step_count"
                "unit": "count",
                "source": null
            },
            "timestamp": 1646573476438
}
```

**Comment by divyav2020** [comment]
@avaidyam 
This issue is fixed in iOS. But android build is not stable now. So we will update the status once the android build is stable

**Comment by ZCOEngineer** [comment]
Testing Android with new build today

**Comment by divyav2020** [comment]
This issue is fixed in iOS and Android

**Comment by dcurrey88** [comment]
This does not appear to be fixed in staging. The four keys specified [value, type, unit, source] are not present. 

For android, sometimes "source" is missing:
```
{'data': {'type': 'speed', 'unit': 'meters per second', 'value': 0.9995065},
  'sensor': 'lamp.steps',
  'timestamp': 1649783955264},
 {'data': {'source': 'com.google.android.gms',
   'type': 'active_minutes',
   'unit': 'minutes',
   'value': 1},
  'sensor': 'lamp.steps',
  'timestamp': 1649783955264}
```

For iOS, only the "value" field is present:
```
{'sensor': 'lamp.steps',
   'data': {'distance': 3344.070000003092,
    'floors_descended': 9,
    'event': 'resume',
    'active_pace': 0.9602527290837306,
    'floors_ascended': 11,
    'value': 4581,
    'pace': 0,
    'cadence': 0},
   'timestamp': 1649788125486},
```

**Comment by saranya-sajeev** [comment]
@dcurrey88 We have checked current Android beta build and got steps data as below. If the source field is not available from Google fit we will be sending null as below.  Are you using the latest beta build? 
 {
      "data": {
        "source": "com.google.android.gms",
        "type": "step_count",
        "unit": "count",
        "value": 13
      },
      "sensor": "lamp.steps",
      "timestamp": 1649842803515
    },
    {
      "data": {
        "source": null,
        "type": "speed",
        "unit": "meters per second",
        "value": 0.30389535
      },
      "sensor": "lamp.steps",
      "timestamp": 1649842803400
    },
    {
      "data": {
        **"source": null,**
        "type": "distance",
        "unit": "meter",
        "value": 72.93489
      },
      "sensor": "lamp.steps",
      "timestamp": 1649842803399
    },

**Comment by jijopulikkottil** [comment]
@avaidyam Could you please release new TestFlight version, so we can make sure all are testing with latest version of iOS app. Otherwise we have to check the build number is 343 or greater.

**Comment by dcurrey88** [comment]
I am testing on build 352 and it looks like iOS data is in the correct format. But source is None, is this intended behavior?

```
{'sensor': 'lamp.steps',
  'data': {'value': 1687,
   'source': None,
   'type': 'step_count',
   'unit': 'count'},
  'timestamp': 1649963943984},
```
We are still testing the Android version.

**Comment by jijopulikkottil** [comment]
@dcurrey88  Expected behaviour is `'source': null,`. Could you share the user id to check!

**Comment by lukeoftheshire** [comment]
@avaidyam @jijopulikkottil  for this specifically null/None issue specifically - correct me if I'm wrong but I think this may just be a difference between how JS and Python handle null data - so this may not be a *bug* per se.

**Comment by lukeoftheshire** [comment]
Additionally, there are some oddities happening with step data. Per issue https://github.com/BIDMCDigitalPsychiatry/LAMP-platform/issues/133 it looks like we set a goal of having step counts in the db reflect steps taken since the last query (i.e. 10 steps taken between 8:50 and 8:55). However, during testing the following sensor events were generated:

```
{'sensor': 'lamp.steps',
   'data': {'value': 3325,
    'source': None,
    'type': 'step_count',
    'unit': 'count'},
   'timestamp': 1649969962378},
```
and
```
{'sensor': 'lamp.steps',
   'data': {'value': 3292,
    'source': None,
    'type': 'step_count',
    'unit': 'count'},
   'timestamp': 1649969935440},
```

When the timestamps are converted, these occur about ~30 seconds apart - it is implausible that the tester took >3000 steps in a few seconds, so it seems that this is almost certainly more data counted cumulatively.

Interestingly, this data is also present when you select for sensor events where the source is not null:
```
[{'sensor': 'lamp.steps',
  'data': {'value': 707,
   'source': 'com.apple.health.804B093D-9602-4CB3-B2A6-1E34F5325DFC',
   'type': 'step_count',
   'unit': 'count'},
  'timestamp': 1649969959810},
 {'sensor': 'lamp.steps',
  'data': {'value': 220,
   'source': 'com.apple.health.804B093D-9602-4CB3-B2A6-1E34F5325DFC',
   'type': 'step_count',
   'unit': 'count'},
  'timestamp': 1649968856582},
 {'sensor': 'lamp.steps',
  'data': {'value': 711,
   'source': 'com.apple.health.804B093D-9602-4CB3-B2A6-1E34F5325DFC',
   'type': 'step_count',
   'unit': 'count'},
  'timestamp': 1649968348681}]
```

These look to be reasonable estimates of steps taken since previous timestamps - the first one, for instance, suggests 700 steps taken over about 15-20 minutes, which is very reasonable.

Is it possible there are two pieces of code submitting stepcount sensor events? I think we probably don't need both if that's the case - the data with a source is a good solution to issue #133 and the 'cumulative' data is less usable than the other kind anyway.

**Comment by avaidyam** [comment]
> I think this may just be a difference between how JS and Python handle null data - so this may not be a *bug* per se.

We should still be seeing `None` instead of `null` though?

**Comment by lukeoftheshire** [comment]
We are seeing `None` in python - so that's expected.

**Comment by jijopulikkottil** [comment]
> Looks like `lamp.steps` for android is not cumulative, which is good!

In iOS, when fetching steps from pedometer (`source` will be null), we are fetching daily count.
But when fetch data using HealthKit, it will not be cumulative, and the `source` will not be null.
